specs/core/0_beacon-chain.md

```diff
@@ -1772,135 +1773,310 @@ For each `transfer` in `block.body.transfers`:
+        a.data.epoch_boundary_root == get_block_root(state, get_epoch_start_slot(get_current_epoch(state)))
+    ]
+```
 
-Define the following helpers to process attestation inclusion rewards and inclusion distance reward/penalty. For every attestation `a` in `previous_epoch_attestations`:
+```python
+def get_previous_epoch_boundary_attestations(state: BeaconState) -> List[PendingAttestation]:
+    return [
+        a for a in get_previous_attestations(state) if 
+        a.data.epoch_boundary_root == get_block_root(state, get_epoch_start_slot(get_previous_epoch(state)))
+    ]
+```
 
-* Let `inclusion_slot(state, index) = a.inclusion_slot` for the attestation `a` where `index` is in `get_attestation_participants(state, a.data, a.aggregation_bitfield)`. If multiple attestations are applicable, the attestation with lowest `inclusion_slot` is considered.
-* Let `inclusion_distance(state, index) = a.inclusion_slot - a.data.slot` where `a` is the above attestation.
+```python
+def get_previous_epoch_matching_head_attestations(state: BeaconState) -> List[PendingAttestation]:
+    return [
+        a for a in get_previous_epoch_attestations(state) if
+        a.data.beacon_block_root == get_block_root(state, a.data.slot)
+    ]
+```
 
-#### Eth1 data
+**Note**: Total balances computed for the previous epoch might be marginally different than the actual total balances during the previous epoch transition. Due to the tight bound on validator churn each epoch and small per-epoch rewards/penalties, the potential balance difference is very low and only marginally affects consensus safety.
 
-If `next_epoch % EPOCHS_PER_ETH1_VOTING_PERIOD == 0`:
+```python
+def get_winning_root_and_participants(state: BeaconState, shard: Shard) -> Tuple[Bytes32, List[ValidatorIndex]]:
+    all_attestations = get_current_attestations(state) + get_previous_attestations(state)
+    valid_attestations = [
+        a for a in all_attestations if a.data.latest_crosslink == state.latest_crosslinks[shard]
+    ]
+    all_roots = [a.data.crosslink_data_root for a in valid_attestations]
+    
+    def get_attestations_for(root) -> List[PendingAttestation]:
+        return [a for a in valid_attestations if a.data.crosslink_data_root == root]
+        
+    winning_root = max(all_roots, key=lambda r: get_attesting_balance(state, get_attestations_for(r)))
+    
+    return winning_root, get_attesting_indices(get_attestations_for(winning_root))
+```
 
-* If `eth1_data_vote.vote_count * 2 > EPOCHS_PER_ETH1_VOTING_PERIOD * SLOTS_PER_EPOCH` for some `eth1_data_vote` in `state.eth1_data_votes` (ie. more than half the votes in this voting period were for that value), set `state.latest_eth1_data = eth1_data_vote.eth1_data`.
-* Set `state.eth1_data_votes = []`.
+```python
+def earliest_attestation(state: BeaconState, validator_index: ValidatorIndex) -> PendingAttestation:
+    return min([
+        a for a in state.previous_epoch_attestations if
```

Comment:
```suggestion
        a for a in state.latest_attestations if
```

Reply:
Maybe we should just move the logic from #697 into this PR? Would remove the need for the get_current/previous_epoch_attestations functions and clean a lot of this up, and if we're going to pythonize all the changes from #697 will need to be translated anyway.

Thoughts @JustinDrake @djrtwo?